platform/darwin/src/MGLOfflineTask.mm

```diff
@@ -0,0 +1,189 @@
+    [NSException raise:@"Method unavailable"
+                format:
+     @"-[MGLOfflineTask init] is unavailable. "
+     @"Use +[MGLOfflineStorage addTaskForRegion:withContext:completionHandler:] instead."];
+    return nil;
+}
+
+- (instancetype)initWithMBGLRegion:(mbgl::OfflineRegion *)region {
+    if (self = [super init]) {
+        _mbglOfflineRegion = region;
+        _state = MGLOfflineTaskStateInactive;
+        _mbglOfflineRegionObserver = new MBGLOfflineRegionObserver(self);
+        
+        mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+        mbglFileSource->setOfflineRegionObserver(*_mbglOfflineRegion, std::make_unique<MBGLOfflineRegionObserver>(*_mbglOfflineRegionObserver));
+    }
+    return self;
+}
+
+- (void)dealloc {
+    delete _mbglOfflineRegionObserver;
+    _mbglOfflineRegionObserver = nullptr;
+}
+
+- (id <MGLOfflineRegion>)region {
+    MGLAssertOfflineTaskIsValid();
+    
+    const mbgl::OfflineRegionDefinition &regionDefinition = _mbglOfflineRegion->getDefinition();
+    NSAssert([MGLTilePyramidOfflineRegion conformsToProtocol:@protocol(MGLOfflineRegion_Private)], @"MGLTilePyramidOfflineRegion should conform to MGLOfflineRegion_Private.");
+    return [(id <MGLOfflineRegion_Private>)[MGLTilePyramidOfflineRegion alloc] initWithOfflineRegionDefinition:regionDefinition];
+}
+
+- (NSData *)context {
+    MGLAssertOfflineTaskIsValid();
+    
+    const mbgl::OfflineRegionMetadata &metadata = _mbglOfflineRegion->getMetadata();
+    return [NSData dataWithBytes:&metadata[0] length:metadata.size()];
+}
+
+- (void)resume {
+    MGLAssertOfflineTaskIsValid();
+    
+    mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+    mbglFileSource->setOfflineRegionDownloadState(*_mbglOfflineRegion, mbgl::OfflineRegionDownloadState::Active);
+    self.state = MGLOfflineTaskStateActive;
+}
+
+- (void)suspend {
+    MGLAssertOfflineTaskIsValid();
+    
+    mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+    mbglFileSource->setOfflineRegionDownloadState(*_mbglOfflineRegion, mbgl::OfflineRegionDownloadState::Inactive);
+    self.state = MGLOfflineTaskStateInactive;
+}
+
+- (void)invalidate {
+    MGLAssertOfflineTaskIsValid();
+    
+    self.state = MGLOfflineTaskStateInvalid;
+}
+
+- (void)requestProgress {
+    MGLAssertOfflineTaskIsValid();
+    
+    mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+    
+    __weak MGLOfflineTask *weakSelf = self;
+    mbglFileSource->getOfflineRegionStatus(*_mbglOfflineRegion, [&, weakSelf](__unused std::exception_ptr exception, mbgl::optional<mbgl::OfflineRegionStatus> status) {
+        if (status) {
+            mbgl::OfflineRegionStatus checkedStatus = *status;
+            dispatch_async(dispatch_get_main_queue(), ^{
+                MGLOfflineTask *strongSelf = weakSelf;
```

Comment:
The weak reference is either valid or `nil`, in which case the following line is a no-op.

Or are you referring to the fact that I skipped doing a strong/weak dance directly inside the outer lambda? I might be mistaken, but I don’t think an extra level of weak indirection is required.